File: lab8/birthday.py
    
    

import sys
import string
import base64

import random
import string

from Crypto.Hash import SHA256

def raw2hex(raw):
    return raw.encode('hex')


def hex2raw(hexstring):
    return base64.b16decode(hexstring)

hexdigits = '0123456789ABCDEF'

def hash(message):
    h = SHA256.new()
    h.update(message)
    return h.digest()


def get_random_message():
    message = ''
    for _ in range(0, random.randint(4, 100)):
        if (random.choice([0, 1]) == 1):
            message += str(random.choice(string.digits))
        else:
            message += random.choice(string.letters)
    return message, message[0:4]

def main():
    # Try to find a collision on the first 4 bytes (32 bits)
    
    # Step 1. Generate 2^16 different random messages
    random_messages = []
    random_small_messages = []
    hashDict = {}
    for _ in range(pow(2,16)):
        m,smallm = get_random_message()
        random_messages.append(m)
        random_small_messages.append(smallm)
    
    # Step 2. Compute hashes
    random_hashes = []
    for _ in range(pow(2, 16)):
        m = random_small_messages[_]
        mhash = hash(m)
        random_hashes.append(mhash)

        if (mhash in hashDict and (hashDict[mhash])):
            hashDict[mhash].append(random_messages[_])
        else:
            hashDict[mhash] = [random_messages[_]]


    # Step 3. Check if there exist two hashes that match in the first
    # four bytes.
    
    # Step 3a. If a match is found, print the messages and hashes
    
    # Step 3b. If no match is found, repeat the attack with a new set
    # of random messages
    '''
    for i in range(0, pow(2, 16)):
        for j in range(0, pow(2, 16)):
            if i == j:
                continue
            if (random_messages[i] != random_messages[j]) and (random_hashes[i] == random_hashes[j]):
                print('Message #1: ' + random_messages[i])
                print('Message #2:' + random_messages[j])
    '''
    for key in hashDict:
        if len(hashDict[key]) > 1:
            print('key:' + key)
            print(hashDict[key])

if __name__ == "__main__":
    main()

File: lab9/birthday.c
#include <openssl/sha.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

/* We want a collision in the first 4 bytes = 2^16 attempts */
#define N_BITS  16

int raw2int4(unsigned char * digest) {
    int i;
    int sum = 0;

    for (i = 0; i < 3; i++) {
        sum += sum * 256 + digest[i];
    }

    return sum;
}

void hexdump(unsigned char * string, int length) {
    int i;
    for (i = 0; i < length; i++) {
        printf("%02x", string[i]);
    }
}

int main(int argc, char * argv[]) {
    //uint32_t attempt;     /* Iterate through 16 bits of the 32; use the rest to run different attacks */
    unsigned char md[20]; /* SHA-1 outputs 160-bit digests */

    /* Try to find a collision on the first 4 bytes (32 bits) */
    unsigned int number_of_tries = 2 << N_BITS;
    unsigned int i, j;

    char ** random_messages = (char**) malloc(number_of_tries * sizeof(char *));
    int * random_hashes = (int *) calloc(number_of_tries, sizeof(int));

    /* Step 1. Generate 2^16 different random messages */
    for (int i = 0; i < number_of_tries; i++)
    {
        char * m = malloc(16 * sizeof(char));
        random_messages[i] = m;
    }

    /* Step 2. Compute hashes */
    SHA_CTX context;
    SHA1_Init(&context);
    for (int i = 0; i < number_of_tries; i++)
    {
        char *m = random_messages[i];

        SHA1_Update(&context, m, 16);
        SHA1_Final(md, &context); /* md must point to at least 20 bytes of valid memory */

        random_hashes[i] = raw2int4(md);
    }

    for (i = 0; i < number_of_tries; i++)
    {
        for (j = i + 1; j < number_of_tries; j++)
        {
            /* Step 3. Check if there exist two hashes that match in the first four bytes */
            if ((memcmp(random_messages[i], random_messages[j], 4) != 0) && (random_hashes[i] == random_hashes[j]))
            {
                /* Step 3a. If a match is found, print the messages and hashes */
                printf("%s has the same has as %s \n", random_messages[i], random_messages[j]);
            }
        }
    }

    return 0;
}
